scripts/staging/tSNE.dml

```diff
@@ -0,0 +1,156 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+# This script performs dimensionality reduction using tSNE algorithm based on
+# the paper: Visualizing Data using t-SNE, Maaten et. al.
+
+#X = read($X)
+
+X = read("hdfs://rr-dense1.softlayer.com/user/iyounus/data/mnist_train.csv")
+#X = read("hdfs://rr-dense1.softlayer.com/user/iyounus/data/mnist_train.csv")
+
+# X = read("file:///home/iyounus/MNIST/mnist_train.csv")
+# X = read("hdfs:///user/iyounus/data/mnist_test_no_labels.csv")
+# X = read("hdfs:///user/iyounus/data/mnist2500_X.csv")
+
+
+ndim = ifdef($ndim, 2)
+perplexity = ifdef($perplexity, 30)
+lr = ifdef($rate, 300.)
+momentum = ifdef($momentum, 0.9)
+iter = ifdef($iterations, 1000)
+
+print("starting....")
+
+distance_matrix = function(matrix[double] X)
+  return (matrix[double] out) {
+    n = nrow(X)
+    s = rowSums(X * X)
+    out = - 2*X %*% t(X) + s + t(s)
+  }
+
+
+x2p = function(matrix[double] X, double perplexity)
+return(matrix[double] P) {
+  print("x2p....")
+  tol = 1.0e-5
+  INF = 1.0e20
+  n = nrow(X)
+  print(n)
+  D = distance_matrix(X)
+
+  P = matrix(0, rows=n, cols=n)
+  beta = matrix(1, rows=n, cols=1)
+  betamax = matrix(INF, rows=n, cols=1)
+  betamin = matrix(0, rows=n, cols=1)
+  Hdiff = matrix(INF, rows=n, cols=1)
+  logU = log(perplexity)
+
+  ZERODIAG = (diag(matrix(-1, rows=n, cols=1)) + 1)
+  itr = 1
+  while (mean(abs(Hdiff)) > tol & itr < 50) {
+    P = exp(-D * beta)
+    P = P * ZERODIAG
+    sum_Pi = rowSums(P)
+    W = rowSums(P * D)
+    Ws = W/sum_Pi
+    H = log(sum_Pi) + beta * Ws
+    P = P/sum_Pi
+    Hdiff = H - logU
+
+    for(i in 1:n) {
```

Comment:
@iyounus I tried to matricize this for loop using following DML:
```python
betamin = betamin + (Hdiff > 0)*beta
betamax = betamax + (Hdiff <= 0)*beta
beta = 2*(Hdiff > 0)*(betamax == INF)*beta + (Hdiff > 0)*(betamax != INF)*(beta - betamax)/2 + (Hdiff <= 0)*(betamin == 0)*beta/2 + (Hdiff <= 0)*(betamin != 0)*(beta + betamin)/2
```
I wrote a sample DML to test the performance and also accuracy:
```python
n = 100000
INF = 1.0e20
betamax = matrix(INF, rows=n, cols=1)
betamin = matrix(0, rows=n, cols=1)
beta = matrix(seq(1, n), rows=n, cols=1)
Hdiff = matrix(seq(1, n)-(n/2), rows=n, cols=1)
# INSERT for loop or above expression here
print(toString(beta[10:20,]))
print(sum(beta))
```

Based on initial comparison, the for loop version took 27 seconds and the matricized version took 0.226 seconds. 

Not saying that this is the bottleneck, but thought my suggestion could help :)